art/estimators/certification/deep_z/pytorch.py

```diff
@@ -0,0 +1,215 @@
+        :param preprocessing_defences: Preprocessing defence(s) to be applied by the classifier.
+        :param postprocessing_defences: Postprocessing defence(s) to be applied by the classifier.
+        :param preprocessing: Tuple of the form `(subtrahend, divisor)` of floats or `np.ndarray` of values to be
+               used for data preprocessing. The first value will be subtracted from the input. The input will then
+               be divided by the second one.
+        :param device_type: Type of device on which the classifier is run, either `gpu` or `cpu`.
+        """
+
+        warnings.warn(
+            "\nThis estimator does not support networks which have dense layers before convolutional. "
+            "We currently infer a reshape when a neural network goes from convolutional layers to "
+            "dense layers. If your use case does not fall into this pattern then consider "
+            "directly building a certifier network with the "
+            "custom layers found in art.estimators.certification.deepz.deep_z.py\n"
+        )
+
+        super().__init__(
+            model=model,
+            loss=loss,
+            input_shape=input_shape,
+            nb_classes=nb_classes,
+            optimizer=optimizer,
+            channels_first=channels_first,
+            clip_values=clip_values,
+            preprocessing_defences=preprocessing_defences,
+            postprocessing_defences=postprocessing_defences,
+            preprocessing=preprocessing,
+            device_type=device_type,
+        )
+
+        modules = []
+
+        # pylint: disable=W0613
+        def forward_hook(input_module, hook_input, hook_output):
+            modules.append(input_module)
+
+        for module in model.children():
+            module.register_forward_hook(forward_hook)
+
+        if self.channels_first:
+            input_for_hook = torch.rand(self.input_shape).to(self.device)
+        else:
+            raise ValueError("Please provide data in channels first format")
+
+        input_for_hook = torch.unsqueeze(input_for_hook, dim=0)
+        model(input_for_hook)  # hooks are fired sequentially from model input to the output
+
+        self.ops = nn.ModuleList()
+        for module in modules:
+            print("registered", type(module))
+            if isinstance(module, torch.nn.modules.conv.Conv2d):
+                zono_conv = ZonoConv(
+                    in_channels=module.in_channels,
+                    out_channels=module.out_channels,
+                    kernel_size=module.kernel_size,  # type: ignore
+                    stride=module.stride,  # type: ignore
+                    dilation=module.dilation,  # type: ignore
+                    padding=module.padding,  # type: ignore
+                )
+                zono_conv.conv.weight.data = module.weight.data.to(self.device)
+                zono_conv.bias.data = module.bias.data.to(self.device)  # type: ignore
```

Comment:
Why are the type ignores required?

Reply:
Ignored due to `nn.Conv2d` requiring the type hinting indicated in the init of `ZonoConv(torch.nn.Module)`. However, reading layer information using `module.X` can return types of a slightly different form (eg `Tuple[int, ...]` vs `Tuple[int]`) which nonetheless both function when passed into `nn.Conv2d`. Hence, using an ignore to suppress the warning.